File: 02-arrays-and-linked-lists/03-singly-linked-list/javascript/singlyLinkedListDemo.test.js
// 03 單向鏈結串列測試（JavaScript）/ Tests for singly linked list (JavaScript).  // Bilingual file header.

"use strict";  // Enable strict mode for tests as well.  

const test = require("node:test");  // Use Node's built-in test runner.  
const assert = require("node:assert/strict");  // Use strict assertions.  

const {  // Import code under test.  
    SinglyLinkedList,  // Singly linked list implementation.  
    buildOrderedList,  // Helper that builds [0..n-1].  
} = require("./singlyLinkedListDemo.js");  // Load module relative to this file.  

test("pushFront hops is always 0", () => {  // pushFront should never traverse next pointers.  
    const lst = new SinglyLinkedList();  // Start with empty list.  
    for (const v of [1, 2, 3, 4]) {  // Insert a few values.  
        const cost = lst.pushFront(v);  // Insert at head.  
        assert.equal(cost.hops, 0);  // Validate cost.  
    }  // Close loop.  
});  // End test.  

test("pushBack hops equals max(0, n-1)", () => {  // pushBack should traverse to tail without a tail pointer.  
    for (const n of [0, 1, 2, 3, 5, 10]) {  // Multiple sizes including boundary cases.  
        const lst = buildOrderedList(n);  // Build [0..n-1].  
        const cost = lst.pushBack(999);  // Append sentinel.  
        const expected = n === 0 ? 0 : n - 1;  // Expected hops.  
        assert.equal(cost.hops, expected);  // Validate traversal cost.  
        assert.equal(lst.size, n + 1);  // Validate size update.  
        assert.equal(lst.toArray()[lst.size - 1], 999);  // Validate tail value.  
    }  // Close loop.  
});  // End test.  

test("getWithCost(i) returns value i with hops i on ordered list", () => {  // getWithCost should traverse i links.  
    const lst = buildOrderedList(10);  // Build [0..9].  
    for (const i of [0, 1, 5, 9]) {  // Sample indices.  
        const r = lst.getWithCost(i);  // Fetch value + cost.  
        assert.equal(r.value, i);  // Value should equal index in ordered list.  
        assert.equal(r.cost.hops, i);  // Hops should equal index.  
    }  // Close loop.  
});  // End test.  

test("insertAt cost and ordering", () => {  // insertAt should splice correctly and report hops.  
    const a = buildOrderedList(5);  // Build [0..4].  
    const c0 = a.insertAt(0, 99);  // Insert at head.  
    assert.equal(c0.hops, 0);  // Head insertion has zero hops.  
    assert.deepEqual(a.toArray(), [99, 0, 1, 2, 3, 4]);  // Validate ordering.  

    const b = buildOrderedList(5);  // Build [0..4] again.  
    const c3 = b.insertAt(3, 77);  // Insert at index 3.  
    assert.equal(c3.hops, 2);  // Predecessor is index 2 => 2 hops.  
    assert.deepEqual(b.toArray(), [0, 1, 2, 77, 3, 4]);  // Validate ordering.  
});  // End test.  

test("removeAt cost and ordering", () => {  // removeAt should bypass node correctly and report hops.  
    const a = buildOrderedList(5);  // Build [0..4].  
    const r0 = a.removeAt(0);  // Remove head.  
    assert.equal(r0.value, 0);  // Removed value should be 0.  
    assert.equal(r0.cost.hops, 0);  // Head removal has zero hops.  
    assert.deepEqual(a.toArray(), [1, 2, 3, 4]);  // Validate ordering.  

    const b = buildOrderedList(5);  // Build [0..4] again.  
    const r3 = b.removeAt(3);  // Remove value 3.  
    assert.equal(r3.value, 3);  // Removed value should be 3.  
    assert.equal(r3.cost.hops, 2);  // Predecessor is index 2 => 2 hops.  
    assert.deepEqual(b.toArray(), [0, 1, 2, 4]);  // Validate ordering.  
});  // End test.  

test("indexOfWithCost hops", () => {  // indexOfWithCost should count next traversals.  
    const lst = buildOrderedList(5);  // Build [0..4].  
    const f0 = lst.indexOfWithCost(0);  // Find head value.  
    assert.equal(f0.index, 0);  // Index should be 0.  
    assert.equal(f0.cost.hops, 0);  // Finding head needs 0 hops.  
    const f3 = lst.indexOfWithCost(3);  // Find value at index 3.  
    assert.equal(f3.index, 3);  // Index should be 3.  
    assert.equal(f3.cost.hops, 3);  // Reaching index 3 needs 3 hops.  
    const fn = lst.indexOfWithCost(999);  // Search for missing value.  
    assert.equal(fn.index, -1);  // Not found sentinel.  
    assert.equal(fn.cost.hops, 5);  // Not found should traverse n hops.  
});  // End test.  

test("invalid indices throw", () => {  // Methods should validate indices.  
    const lst = buildOrderedList(3);  // Build [0..2].  
    assert.throws(() => lst.getWithCost(-1), RangeError);  // Invalid get.  
    assert.throws(() => lst.getWithCost(3), RangeError);  // Invalid get.  
    assert.throws(() => lst.insertAt(4, 0), RangeError);  // Invalid insert.  
    assert.throws(() => lst.removeAt(3), RangeError);  // Invalid remove.  
});  // End test.  

File: 02-arrays-and-linked-lists/03-singly-linked-list/java/SinglyLinkedListDemoTest.java
// 03 單向鏈結串列測試（Java）/ Tests for singly linked list (Java).  // Bilingual file header.

import java.util.Arrays;  // Use Arrays.equals for comparing expected arrays.  

public final class SinglyLinkedListDemoTest {  // Minimal test runner without JUnit/Maven.  
    private SinglyLinkedListDemoTest() {  // Prevent instantiation.  
        // No instances.  // Clarify intent.  
    }  // Close constructor.  

    private static void assertTrue(boolean condition, String message) {  // Minimal boolean assertion helper.  
        if (!condition) {  // Fail when condition is false.  
            throw new AssertionError("FAIL: " + message);  // Throw to signal test failure.  
        }  // Close failure branch.  
    }  // End assertTrue.  

    private static void assertEquals(long expected, long actual, String message) {  // Minimal long equality assertion helper.  
        if (expected != actual) {  // Fail when values differ.  
            throw new AssertionError("FAIL: " + message + " (expected=" + expected + ", actual=" + actual + ")");  // Throw mismatch.  
        }  // Close failure branch.  
    }  // End assertEquals.  

    private static void assertArrayEquals(int[] expected, int[] actual, String message) {  // Minimal array equality helper.  
        if (!Arrays.equals(expected, actual)) {  // Fail when arrays differ.  
            throw new AssertionError("FAIL: " + message + " (expected=" + Arrays.toString(expected) + ", actual=" + Arrays.toString(actual) + ")");  // Throw mismatch.  
        }  // Close failure branch.  
    }  // End assertArrayEquals.  

    private static void assertThrows(Class<? extends Throwable> expectedType, Runnable action, String message) {  // Assert that an action throws.  
        try {  // Execute action and expect it to throw.  
            action.run();  // Run action.  
        } catch (Throwable t) {  // Catch any throwable.  
            if (expectedType.isInstance(t)) {  // Accept expected type.  
                return;  // Test passed.  
            }  // Close type check.  
            throw new AssertionError("FAIL: " + message + " (threw " + t.getClass().getName() + ")");  // Wrong type.  
        }  // Close catch.  
        throw new AssertionError("FAIL: " + message + " (no exception thrown)");  // Fail if nothing thrown.  
    }  // End assertThrows.  

    private static void testPushFrontHopsIsZero() {  // pushFront should never traverse next pointers.  
        SinglyLinkedListDemo.SinglyLinkedList lst = new SinglyLinkedListDemo.SinglyLinkedList();  // Create empty list.  
        int[] values = new int[] {1, 2, 3, 4};  // Values to insert.  
        for (int v : values) {  // Iterate values.  
            SinglyLinkedListDemo.OperationCost cost = lst.pushFront(v);  // Insert at head.  
            assertEquals(0, cost.hops, "pushFront hops should be 0");  // Validate cost.  
        }  // Close loop.  
    }  // End testPushFrontHopsIsZero.  

    private static void testPushBackHopsEqualsSizeMinusOne() {  // pushBack should traverse to tail without a tail pointer.  
        int[] ns = new int[] {0, 1, 2, 3, 5, 10};  // Multiple sizes.  
        for (int n : ns) {  // Iterate test cases.  
            SinglyLinkedListDemo.SinglyLinkedList lst = SinglyLinkedListDemo.buildOrderedList(n);  // Build [0..n-1].  
            SinglyLinkedListDemo.OperationCost cost = lst.pushBack(999);  // Append sentinel.  
            int expected = (n == 0) ? 0 : (n - 1);  // Expected hops for head-only list.  
            assertEquals(expected, cost.hops, "pushBack hops should equal max(0,n-1)");  // Validate traversal cost.  
            assertEquals(n + 1, lst.size(), "size should increase by one");  // Validate size update.  
            int[] arr = lst.toArray();  // Convert to array for tail check.  
            assertEquals(999, arr[arr.length - 1], "pushBack should append to tail");  // Validate tail value.  
        }  // Close loop.  
    }  // End testPushBackHopsEqualsSizeMinusOne.  

    private static void testGetWithCostHopsEqualsIndex() {  // getWithCost(i) should traverse i links.  
        SinglyLinkedListDemo.SinglyLinkedList lst = SinglyLinkedListDemo.buildOrderedList(10);  // Build [0..9].  
        int[] indices = new int[] {0, 1, 5, 9};  // Sample indices.  
        for (int i : indices) {  // Iterate indices.  
            SinglyLinkedListDemo.GetResult r = lst.getWithCost(i);  // Fetch value + cost.  
            assertEquals(i, r.value, "value should equal index for ordered list");  // Validate value.  
            assertEquals(i, r.cost.hops, "getWithCost hops should equal index");  // Validate cost.  
        }  // Close loop.  
    }  // End testGetWithCostHopsEqualsIndex.  

    private static void testInsertAtHopsEqualsIndexMinusOne() {  // insertAt(i) should traverse (i-1) links for i>0.  
        SinglyLinkedListDemo.SinglyLinkedList a = SinglyLinkedListDemo.buildOrderedList(5);  // Build [0..4].  
        SinglyLinkedListDemo.OperationCost c0 = a.insertAt(0, 99);  // Insert at head.  
        assertEquals(0, c0.hops, "insertAt(0) hops should be 0");  // Validate cost.  
        assertArrayEquals(new int[] {99, 0, 1, 2, 3, 4}, a.toArray(), "insertAt(0) should insert at head");  // Validate order.  

        SinglyLinkedListDemo.SinglyLinkedList b = SinglyLinkedListDemo.buildOrderedList(5);  // Build [0..4] again.  
        SinglyLinkedListDemo.OperationCost c3 = b.insertAt(3, 77);  // Insert at index 3.  
        assertEquals(2, c3.hops, "insertAt(3) hops should be 2");  // Validate cost.  
        assertArrayEquals(new int[] {0, 1, 2, 77, 3, 4}, b.toArray(), "insertAt should splice correctly");  // Validate order.  
    }  // End testInsertAtHopsEqualsIndexMinusOne.  

    private static void testRemoveAtHopsEqualsIndexMinusOne() {  // removeAt(i) should traverse (i-1) links for i>0.  
        SinglyLinkedListDemo.SinglyLinkedList a = SinglyLinkedListDemo.buildOrderedList(5);  // Build [0..4].  
        SinglyLinkedListDemo.RemoveResult r0 = a.removeAt(0);  // Remove head.  
        assertEquals(0, r0.value, "removeAt(0) should remove head value");  // Validate removed value.  
        assertEquals(0, r0.cost.hops, "removeAt(0) hops should be 0");  // Validate cost.  
        assertArrayEquals(new int[] {1, 2, 3, 4}, a.toArray(), "removeAt(0) should remove head");  // Validate order.  

        SinglyLinkedListDemo.SinglyLinkedList b = SinglyLinkedListDemo.buildOrderedList(5);  // Build [0..4] again.  
        SinglyLinkedListDemo.RemoveResult r3 = b.removeAt(3);  // Remove value 3.  
        assertEquals(3, r3.value, "removeAt(3) should remove value 3");  // Validate removed value.  
        assertEquals(2, r3.cost.hops, "removeAt(3) hops should be 2");  // Validate cost.  
        assertArrayEquals(new int[] {0, 1, 2, 4}, b.toArray(), "removeAt should bypass removed node");  // Validate order.  
    }  // End testRemoveAtHopsEqualsIndexMinusOne.  

    private static void testIndexOfWithCostHops() {  // indexOfWithCost should count next traversals.  
        SinglyLinkedListDemo.SinglyLinkedList lst = SinglyLinkedListDemo.buildOrderedList(5);  // Build [0..4].  
        SinglyLinkedListDemo.FindResult f0 = lst.indexOfWithCost(0);  // Find head value.  
        assertEquals(0, f0.index, "indexOf(0) should return 0");  // Validate index.  
        assertEquals(0, f0.cost.hops, "finding head needs 0 hops");  // Validate cost.  
        SinglyLinkedListDemo.FindResult f3 = lst.indexOfWithCost(3);  // Find middle value.  
        assertEquals(3, f3.index, "indexOf(3) should return 3");  // Validate index.  
        assertEquals(3, f3.cost.hops, "finding index 3 needs 3 hops");  // Validate cost.  
        SinglyLinkedListDemo.FindResult fn = lst.indexOfWithCost(999);  // Search for missing value.  
        assertEquals(-1, fn.index, "indexOf(missing) should return -1");  // Validate not-found sentinel.  
        assertEquals(5, fn.cost.hops, "missing value should traverse n hops");  // Validate cost.  
    }  // End testIndexOfWithCostHops.  

    private static void testInvalidIndicesThrow() {  // Methods should validate indices.  
        SinglyLinkedListDemo.SinglyLinkedList lst = SinglyLinkedListDemo.buildOrderedList(3);  // Build [0..2].  
        assertThrows(IndexOutOfBoundsException.class, () -> lst.getWithCost(-1), "getWithCost should throw on negative index");  // Invalid get.  
        assertThrows(IndexOutOfBoundsException.class, () -> lst.getWithCost(3), "getWithCost should throw on index==size");  // Invalid get.  
        assertThrows(IndexOutOfBoundsException.class, () -> lst.insertAt(4, 0), "insertAt should throw on index>size");  // Invalid insert.  
        assertThrows(IndexOutOfBoundsException.class, () -> lst.removeAt(3), "removeAt should throw on index==size");  // Invalid remove.  
    }  // End testInvalidIndicesThrow.  

    public static void main(String[] args) {  // Run all tests and print status.  
        testPushFrontHopsIsZero();  // Run pushFront cost test.  
        testPushBackHopsEqualsSizeMinusOne();  // Run pushBack cost test.  
        testGetWithCostHopsEqualsIndex();  // Run get cost test.  
        testInsertAtHopsEqualsIndexMinusOne();  // Run insert cost test.  
        testRemoveAtHopsEqualsIndexMinusOne();  // Run remove cost test.  
        testIndexOfWithCostHops();  // Run find cost test.  
        testInvalidIndicesThrow();  // Run index validation test.  
        System.out.println("All tests PASSED.");  // Print success.  
    }  // End main.  
}  // End SinglyLinkedListDemoTest.  

